shapefile-reader.js

```diff
@@ -19,35 +22,91 @@ function ShapefileStream (files, options) {
   }
   if (!files.shp) throw new Error('No shp property is set')
 
-  this.header = null
+  this.headers = null
 
   // Create readers
+  this._files = files
   this._shpReader = new ShpReader(files.shp)
+  if (files.dbf) this._dbfReader = new DbfReader(files.dbf)
 
   // State
   this._inBody = false
 }
 
+ShapefileStream.prototype._readHeader = function (callback) {
+  const self = this
+
+  // read character encoding from .cpg
+  if (this._files.hasOwnProperty('cpg')) {
+    this._files.cpg.pipe(endpoint(function (err, buffer) {
+      if (err) return callback(err)
+
+      self._dbfReader.setEncoding(buffer.toString('ascii'))
+      phase2()
+    }))
+  } else {
+    phase2()
+  }
+
+  // read header from .shp (contains shapes) and .bdf (contains properties)
+  function phase2 () {
+    // always read .shp
+    const jobs = {
+      shp: (done) => self._shpReader.readHeader(done)
+    }
+
+    // if .dbf exists, read that too
+    if (self._dbfReader) {
+      jobs.bdf = (done) => self._dbfReader.readHeader(done)
+    }
+
+    // read headers
+    async.parallel(jobs, callback)
+  }
+}
+
+ShapefileStream.prototype._readRecord = function (callback) {
+  if (this._dbfReader) {
+    async.parallel({
+      shp: (done) => this._shpReader.readRecord(done),
+      dbf: (done) => this._dbfReader.readRecord(done)
+    }, function (err, record) {
+      if (err) return callback(err)
+
+      if (record.shp === null && record.dbf === null) {
+        callback(null, null)
+      } else if (record.shp !== null && record.dbf !== null) {
+        record.shp.properties = record.dbf
+        callback(null, record.shp)
+      } else {
+        callback(new Error('.shp and .bdf don\'t have the same amount of records'))
+      }
+    })
+  } else {
+    this._shpReader.readRecord(callback)
```

Comment:
I'd like to hear your thoughts on this:

I usually tell others to handle the simplest case first and return early. Here it would be the `else` because it's so short, meaning you'd  `if (!this._dbfReader) return this._shpReader.readRecord(callback)`. What do you think? I usually argue that this makes the code easier to read because you don't have to remember too much when reading the code, but I'd like to hear counter arguments (or reasons it doesn't matter)